### PR TITLE
Improve Molfile IO and validation

### LINKED ISSUE #66: Molfile IO and validation
This is a more targeted issue to continue the discussion we started in #60.

### PR DESCRIPTION
**Pull request by flange-ipb**
closes #66

### CODE REVIEW

**Comment by flange-ipb** [comment]
I have the feeling the io module file is becoming too crowded with different functionalities. Should it be refactored into packages/modules?

**Comment by JanCBrammer** [comment]
> I have the feeling the io module file is becoming too crowded with different functionalities. Should it be refactored into packages/modules?

Which structure do you have in mind?

File: tests/test_io.py [hunk 1]
```diff
@@ -186,3 +224,199 @@ def test_parsing_bond_block():
...
+@pytest.mark.parametrize(
+    "lines, expected_error_msg",
+    [
+        (
+            [
+                "M  V30 abc-",
+                "def",
+            ],
+            'Invalid concatenation of lines "M  V30 abc-" and "def"',
+        ),
```
[the remark this answers is not in the document]
Comment (flange-ipb) [reply tests/test_io.py:337]: This case would not raise an exception and `_concat_lines_with_dash(lines)` returns the same. I will add it as a "positive" test case to `test_concat_lines_with_dash`.

Reason: BIOVIA's specification explains the concatenation rule in the CTAB (V3000) chapter. V3000 CTAB lines must begin with "M  V30 ". It also means that V2000 CTABs do not have the concatenation rule.

[CDK](https://github.com/cdk/cdk/blob/dee3532048201c03da52b2fe27155a18a2bd4a66/storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java#L834) and [rdkit](https://github.com/rdkit/rdkit/blob/a1df50bec8387b33c26bb91bf722902bcf657d47/Code/GraphMol/FileParsers/MolFileParser.cpp#L150) have a similar logic.

**Comment by flange-ipb** [comment]
> Which structure do you have in mind?

- new package (aka path with `__init__.py`) called `io`
- move functions `graph_from_file` and `graph_from_molfile_text` the protected functions they call into `/tucan/io/molfile_reader.py`
- move `graph_from_tucan` into `/tucan/parser/parser.py`
- `/tucan/io.py` is now empty and can be deleted
- a future function `graph_to_molfile` would be in `/tucan/io/molfile_writer.py`

Now, existing imports of functions from `tucan.io` would be broken. I think this can be fixed by arranging the imports in `/tucan/io/__init__.py`.

File: tucan/io.py [hunk 2]
```diff
@@ -51,8 +51,32 @@ def graph_from_tucan(tucan: str) -> nx.Graph:
...
+            else:
+                raise MolfileParserException(
+                    f'Invalid concatenation of lines "{prev_line}" and "{curr_line}"'
+                )
+        else:
+            result.append(curr_line)
+
+    result.append(lines[0])
+    result.reverse()
+    return result
```
Comment (JanCBrammer) [inline tucan/io.py:79]: Maybe we could flatten the logic / remove some nesting?

```Python
def _concat_lines_with_dash(lines: List[str]) -> List[str]:
    final_lines = []
    lines = deque(lines)

    while lines:
        curr_line = lines.popleft()

        if not lines:
            final_lines.append(curr_line)
            break

        if not (curr_line.startswith("M  V30 ") and curr_line.endswith("-")):
            final_lines.append(curr_line)
            continue

        next_line = lines.popleft()
        if not next_line.startswith("M  V30 "):
            raise MolfileParserException(
                f'Invalid concatenation of lines "{curr_line}" and "{next_line}"'
            )

        lines.appendleft(curr_line[0:-1] + next_line[7:])

    return final_lines
```
Using the [deque](https://realpython.com/python-deque/) should also be (marginally) more efficient compared to popping and inserting on a list.

**Comment by JanCBrammer** [comment]
> * new package (aka path with `__init__.py`) called `io`
> * move functions `graph_from_file`, `graph_from_molfile_text` and the protected functions they call into `/tucan/io/molfile_reader.py`
> * move `graph_from_tucan` into `/tucan/parser/parser.py`
> * `/tucan/io.py` is now empty and can be deleted
> * a future function `graph_to_molfile` would be in `/tucan/io/molfile_writer.py`

Looks good to me. Do you want to add these changes here, or open a new PR?

**Comment by flange-ipb** [comment]
> Looks good to me. Do you want to add these changes here, or open a new PR?

I'll do this in a new PR after this one is merged.

**Comment by flange-ipb** [comment]
Your algorithm for `_concat_lines_with_dash` is a bit slower (12%), but this is not the place to look for the best runtime optimization. My algorithm is clearly less understandable and also alters the input list, which is kind of weird.

**Pull request merged**